src/cryptonote_protocol/cryptonote_protocol_handler.inl

```diff
@@ -599,171 +605,57 @@ namespace cryptonote
-          // sending (0...n-1) transactions.
-          // If requested objects is not empty, then we must have asked for 
-          // some missing transacionts, make sure that they're all there.
-          //
-          // Can I safely re-use this field? I think so, but someone check me!
-          if(!context.m_requested_objects.empty()) 
-          {
-            auto req_tx_it = context.m_requested_objects.find(tx_hash);
-            if(req_tx_it == context.m_requested_objects.end())
-            {
-              LOG_ERROR_CCONTEXT
-              (
-                "Peer sent wrong transaction (NOTIFY_NEW_FLUFFY_BLOCK): "
-                << "transaction with id = " << tx_hash << " wasn't requested, "
-                << "dropping connection"
-              );
-              
-              drop_connection(context, false, false);
-              m_core.resume_mine();
-              return 1;
-            }
-            
-            context.m_requested_objects.erase(req_tx_it);
-          }          
-          
-          // we might already have the tx that the peer
-          // sent in our pool, so don't verify again..
-          if(!m_core.pool_has_tx(tx_hash))
-          {
-            MDEBUG("Incoming tx " << tx_hash << " not in pool, adding");
-            cryptonote::tx_verification_context tvc = AUTO_VAL_INIT(tvc);                        
-            if(!m_core.handle_incoming_tx(tx_blob, tvc, relay_method::block, true) || tvc.m_verifivation_failed)
-            {
-              LOG_PRINT_CCONTEXT_L1("Block verification failed: transaction verification failed, dropping connection");
-              drop_connection(context, false, false);
-              m_core.resume_mine();
-              return 1;
-            }
-            
-            //
-            // future todo: 
-            // tx should only not be added to pool if verification failed, but
-            // maybe in the future could not be added for other reasons 
-            // according to monero-moo so keep track of these separately ..
-            //
-          }
+          have_tx.emplace_back(std::move(tx_blob), crypto::null_hash);
+          if (found_in_payload)
+            extra_block_txs.erase(ebt_it);
         }
-        else
+        else if (m_core.get_transactions({tx_hash}, tx_blobs, missed_txs)
+            && tx_blobs.size() == 1
+            && missed_txs.empty()) // or if on-chain
         {
-          LOG_ERROR_CCONTEXT
-          (
-            "sent wrong tx: failed to parse and validate transaction: "
-            << epee::string_tools::buff_to_hex_nodelimer(tx_blob.blob)
-            << ", dropping connection"
-          );
-            
-          drop_connection(context, false, false);
-          m_core.resume_mine();
-          return 1;
+          have_tx.emplace_back(std::move(tx_blobs.front()), crypto::null_hash);
+          if (found_in_payload)
+            extra_block_txs.erase(ebt_it);
```

Comment:
When I benchmarked this PR, I hit segfault here. 

The stack trace was coming from here [1]. I am putting the stack-trace here for you too. Sadly `ebt_it` is optimized out. But if we go down to `erase` and look at the value `__position`: 
```
{<std::__detail::_Node_iterator_base<std::pair<crypto::hash const, std::pair<cryptonote::transaction, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >, true>> = {_M_cur = 0x0}, <No data fields>}
```

I will debug this further in case this information is not enough. 

1. https://github.com/jeffro256/monero/blob/fe370e9c8186cdf2b7a7e89142042e7d921a5a2e/src/p2p/net_node.h#L322C1-L323C1

```
#0  std::_Hashtable<crypto::hash, std::pair<crypto::hash const, std::pair<cryptonote::transaction, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >, std::allocator<std::pair<crypto::hash const, std::pair<cryptonote::transaction, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > >, std::__detail::_Select1st, std::equal_to<crypto::hash>, std::hash<crypto::hash>, std::__detail::_Mod_range_hashing, std::__detail::_Default_ranged_hash, std::__detail::_Prime_rehash_policy, std::__detail::_Hashtable_traits<true, false, true> >::_M_bucket_index (__n=..., this=0x7fb2554e9a10) at /usr/include/c++/13/bits/hashtable.h:793
#1  std::_Hashtable<crypto::hash, std::pair<crypto::hash const, std::pair<cryptonote::transaction, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >, std::allocator<std::pair<crypto::hash const, std::pair<cryptonote::transaction, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > >, std::__detail::_Select1st, std::equal_to<crypto::hash>, std::hash<crypto::hash>, std::__detail::_Mod_range_hashing, std::__detail::_Default_ranged_hash, std::__detail::_Prime_rehash_policy, std::__detail::_Hashtable_traits<true, false, true> >::erase (this=this@entry=0x7fb2554e9a10, __it=..., __it@entry=...)
    at /usr/include/c++/13/bits/hashtable.h:2322
#2  0x0000555555abdd50 in std::_Hashtable<crypto::hash, std::pair<crypto::hash const, std::pair<cryptonote::transaction, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >, std::allocator<std::pair<crypto::hash const, std::pair<cryptonote::transaction, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > >, std::__detail::_Select1st, std::equal_to<crypto::hash>, std::hash<crypto::hash>, std::__detail::_Mod_range_hashing, std::__detail::_Default_ranged_hash, std::__detail::_Prime_rehash_policy, std::__detail::_Hashtable_traits<true, false, true> >::erase (__it=..., this=0x7fb2554e9a10)
    at /usr/include/c++/13/bits/hashtable.h:980
#3  std::unordered_map<crypto::hash, std::pair<cryptonote::transaction, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::hash<crypto::hash>, std::equal_to<crypto::hash>, std::allocator<std::pair<crypto::hash const, std::pair<cryptonote::transaction, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > > >::erase (__position=..., this=0x7fb2554e9a10) at /usr/include/c++/13/bits/unordered_map.h:753
#4  cryptonote::t_cryptonote_protocol_handler<cryptonote::core>::handle_notify_new_fluffy_block (this=this@entry=0x7ffff7ca2010, command=command@entry=2008, arg=..., 
    context=...) at ~/developments/monero/src/cryptonote_protocol/cryptonote_protocol_handler.inl:633
#5  0x00005555557c70a2 in std::__invoke_impl<int, int (cryptonote::t_cryptonote_protocol_handler<cryptonote::core>::*&)(int, epee::misc_utils::struct_init<cryptonote::NOTIFY_NEW_FLUFFY_BLOCK::request_t>&, cryptonote::cryptonote_connection_context&), cryptonote::t_cryptonote_protocol_handler<cryptonote::core>*&, int&, boost::value_initialized<epee::misc_utils::struct_init<cryptonote::NOTIFY_NEW_FLUFFY_BLOCK::request_t> >&, cryptonote::cryptonote_connection_context&> (__t=<optimized out>, __f=<optimized out>)
    at /usr/include/c++/13/bits/invoke.h:71
#6  std::__invoke<int (cryptonote::t_cryptonote_protocol_handler<cryptonote::core>::*&)(int, epee::misc_utils::struct_init<cryptonote::NOTIFY_NEW_FLUFFY_BLOCK::request_t>&, cryptonote::cryptonote_connection_context&), cryptonote::t_cryptonote_protocol_handler<cryptonote::core>*&, int&, boost::value_initialized<epee::misc_utils::struct_init<cryptonote::NOTIFY_NEW_FLUFFY_BLOCK::request_t> >&, cryptonote::cryptonote_connection_context&> (__fn=<optimized out>) at /usr/include/c++/13/bits/invoke.h:96
#7  std::_Bind<int (cryptonote::t_cryptonote_protocol_handler<cryptonote::core>::*(cryptonote::t_cryptonote_protocol_handler<cryptonote::core>*, std::_Placeholder<1>, std::_Placeholder<2>, std::_Placeholder<3>))(int, epee::misc_utils::struct_init<cryptonote::NOTIFY_NEW_FLUFFY_BLOCK::request_t>&, cryptonote::cryptonote_connection_context&)>::__call<int, int&, boost::value_initialized<epee::misc_utils::struct_init<cryptonote::NOTIFY_NEW_FLUFFY_BLOCK::request_t> >&, cryptonote::cryptonote_connection_context&, 0ul, 1ul, 2ul, 3ul>(std::tuple<int&, boost::value_initialized<epee::misc_utils::struct_init<cryptonote::NOTIFY_NEW_FLUFFY_BLOCK::request_t> >&, cryptonote::cryptonote_connection_context&>&&, std::_Index_tuple<0ul, 1ul, 2ul, 3ul>) (__args=..., this=<optimized out>) at /usr/include/c++/13/functional:506
#8  std::_Bind<int (cryptonote::t_cryptonote_protocol_handler<cryptonote::core>::*(cryptonote::t_cryptonote_protocol_handler<cryptonote::core>*, std::_Placeholder<1>, std::_Placeholder<2>, std::_Placeholder<3>))(int, epee::misc_utils::struct_init<cryptonote::NOTIFY_NEW_FLUFFY_BLOCK::request_t>&, cryptonote::cryptonote_connection_context&)>::operator()<int&, boost::value_initialized<epee::misc_utils::struct_init<cryptonote::NOTIFY_NEW_FLUFFY_BLOCK::request_t> >&, cryptonote::cryptonote_connection_context&, int>(int&, boost::value_initialized<epee::misc_utils::struct_init<cryptonote::NOTIFY_NEW_FLUFFY_BLOCK::request_t> >&, cryptonote::cryptonote_connection_context&) (this=<optimized out>)
    at /usr/include/c++/13/functional:591
#9  epee::net_utils::buff_to_t_adapter<cryptonote::t_cryptonote_protocol_handler<cryptonote::core>, epee::misc_utils::struct_init<cryptonote::NOTIFY_NEW_FLUFFY_BLOCK::request_t>, cryptonote::cryptonote_connection_context, std::_Bind<int (cryptonote::t_cryptonote_protocol_handler<cryptonote::core>::*(cryptonote::t_cryptonote_protocol_handler<cryptonote::core>*, std::_Placeholder<1>, std::_Placeholder<2>, std::_Placeholder<3>))(int, epee::misc_utils::struct_init<cryptonote::NOTIFY_NEW_FLUFFY_BLOCK::request_t>&, cryptonote::cryptonote_connection_context&)> >(cryptonote::t_cryptonote_protocol_handler<cryptonote::core>*, int, epee::span<unsigned char const>, std::_Bind<int (cryptonote::t_cryptonote_protocol_handler<cryptonote::core>::*(cryptonote::t_cryptonote_protocol_handler<cryptonote::core>*, std::_Placeholder<1>, std::_Placeholder<2>, std::_Placeholder<3>))(int, epee::misc_utils::struct_init<cryptonote::NOTIFY_NEW_FLUFFY_BLOCK::request_t>&, cryptonote::cryptonote_connection_context&)>, cryptonote::cryptonote_connection_context&) (powner=0x7ffff7ca2010, context=..., cb=..., in_buff=..., command=2008) at ~/developments/monero/contrib/epee/include/storages/levin_abstract_invoke2.h:175
#10 cryptonote::t_cryptonote_protocol_handler<cryptonote::core>::handle_invoke_map<cryptonote::cryptonote_connection_context> (buff_out=..., handled=@0x7fb2554ea3ff: true, 
    context=..., in_buff=..., command=2008, is_notify=true, this=0x7ffff7ca2010) at ~/developments/monero/src/cryptonote_protocol/cryptonote_protocol_handler.h:79
#11 nodetool::node_server<cryptonote::t_cryptonote_protocol_handler<cryptonote::core> >::handle_invoke_map<nodetool::p2p_connection_context_t<cryptonote::cryptonote_connection_context> > (this=<optimized out>, is_notify=is_notify@entry=true, command=2008, in_buff=..., buff_out=..., context=..., handled=@0x7fb2554ea3ff: true)
    at ~/developments/monero/src/p2p/net_node.h:322
#12 0x00005555557c8d5d in nodetool::node_server<cryptonote::t_cryptonote_protocol_handler<cryptonote::core> >::notify (this=<optimized out>, command=<optimized out>, 
    in_buff=..., context=...) at ~/developments/monero/src/p2p/net_node.h:312
#13 0x0000555555aabe60 in epee::levin::async_protocol_handler<nodetool::p2p_connection_context_t<cryptonote::cryptonote_connection_context> >::handle_recv (
    this=this@entry=0x7fb210003570, ptr=ptr@entry=0x7fb210001518, cb=cb@entry=1489)
    at ~/developments/monero/contrib/epee/include/net/levin_protocol_handler_async.h:533
#14 0x0000555555aad73c in epee::net_utils::connection<epee::levin::async_protocol_handler<nodetool::p2p_connection_context_t<cryptonote::cryptonote_connection_context> > >::start_read()::{lambda(boost::system::error_code const&, unsigned long)#1}::operator()(boost::system::error_code const&, unsigned long) const::{lambda()#1}::operator()() const
    (__closure=<synthetic pointer>) at ~/developments/monero/contrib/epee/include/net/abstract_tcp_server2.inl:406
#15 boost::asio::asio_handler_invoke<epee::net_utils::connection<epee::levin::async_protocol_handler<nodetool::p2p_connection_context_t<cryptonote::cryptonote_connection_context> > >::start_read()::{lambda(boost::system::error_code const&, unsigned long)#1}::operator()(boost::system::error_code const&, unsigned long) const::{lambda()#1}>(epee::net_utils::connection<epee::levin::async_protocol_handler<nodetool::p2p_connection_context_t<cryptonote::cryptonote_connection_context> > >::start_read()::{lambda(boost::system::error_code const&, unsigned long)#1}::operator()(boost::system::error_code const&, unsigned long) const::{lambda()#1}&, ...) (function=<synthetic pointer>...)
    at ~/developments/monero/contrib/depends/x86_64-pc-linux-gnu/include/boost/asio/handler_invoke_hook.hpp:69
#16 boost_asio_handler_invoke_helpers::invoke<epee::net_utils::connection<epee::levin::async_protocol_handler<nodetool::p2p_connection_context_t<cryptonote::cryptonote_connection_context> > >::start_read()::{lambda(boost::system::error_code const&, unsigned long)#1}::operator()(boost::system::error_code const&, unsigned long) const::{lambda()#1}, epee::net_utils::connection<epee::levin::async_protocol_handler<nodetool::p2p_connection_context_t<cryptonote::cryptonote_connection_context> > >::start_read()::{lambda(boost::system::error_code const&, unsigned long)#1}::operator()(boost::system::error_code const&, unsigned long) const::{lambda()#1}>(epee::net_utils::connection<epee::levin::async_protocol_handler<nodetool::p2p_connection_context_t<cryptonote::cryptonote_connection_context> > >::start_read()::{lambda(boost::system::error_code const&, unsigned long)#1}::operator()(boost::system::error_code const&, unsigned long) const::{lambda()#1}&, epee::net_utils::connection<epee::levin::async_protocol_handler<nodetool::p2p_connection_context_t<cryptonote::cryptonote_connection_context> > >::start_read()::{lambda(boost::system::error_code const&, unsigned long)#1}::operator()(boost::system::error_code const&, unsigned long) const::{lambda()#1}&) (context=<synthetic pointer>..., function=<synthetic pointer>...)
    at ~/developments/monero/contrib/depends/x86_64-pc-linux-gnu/include/boost/asio/detail/handler_invoke_helpers.hpp:37
#17 boost::asio::detail::completion_handler<epee::net_utils::connection<epee::levin::async_protocol_handler<nodetool::p2p_connection_context_t<cryptonote::cryptonote_connection_context> > >::start_read()::{lambda(boost::system::error_code const&, unsigned long)#1}::operator()(boost::system::error_code const&, unsigned long) const::{lambda()#1}>::do_complete(boost::asio::detail::task_io_service*, boost::asio::detail::task_io_service_operation*, boost::system::error_code const&, unsigned long) (owner=0x555556a96a10, 
    base=0x7fb210018910) at ~/developments/monero/contrib/depends/x86_64-pc-linux-gnu/include/boost/asio/detail/completion_handler.hpp:68
#18 0x0000555555768d65 in boost::asio::detail::task_io_service_operation::complete (bytes_transferred=0, ec=..., owner=..., this=<optimized out>)
    at ~/developments/monero/contrib/depends/x86_64-pc-linux-gnu/include/boost/asio/detail/task_io_service_operation.hpp:38
#19 boost::asio::detail::strand_service::do_complete (owner=0x555556a96a10, base=0x7fb210003c80, ec=...)
    at ~/developments/monero/contrib/depends/x86_64-pc-linux-gnu/include/boost/asio/detail/impl/strand_service.ipp:167
#20 0x0000555555a63c42 in boost::asio::detail::task_io_service_operation::complete (bytes_transferred=<optimized out>, ec=..., owner=..., this=0x7fb210003c80)
    at ~/developments/monero/contrib/depends/x86_64-pc-linux-gnu/include/boost/asio/detail/task_io_service_operation.hpp:38
#21 boost::asio::detail::task_io_service::do_run_one (ec=..., this_thread=..., lock=..., this=0x555556a96a10)
    at ~/developments/monero/contrib/depends/x86_64-pc-linux-gnu/include/boost/asio/detail/impl/task_io_service.ipp:372
#22 boost::asio::detail::task_io_service::run (ec=..., this=0x555556a96a10)
    at ~/developments/monero/contrib/depends/x86_64-pc-linux-gnu/include/boost/asio/detail/impl/task_io_service.ipp:149
#23 boost::asio::io_service::run (this=<optimized out>) at ~/developments/monero/contrib/depends/x86_64-pc-linux-gnu/include/boost/asio/impl/io_service.ipp:59
#24 epee::net_utils::boosted_tcp_server<epee::levin::async_protocol_handler<nodetool::p2p_connection_context_t<cryptonote::cryptonote_connection_context> > >::worker_thread
    (this=0x555556aaf910) at ~/developments/monero/contrib/epee/include/net/abstract_tcp_server2.inl:1317
#25 0x00005555563929f5 in boost::(anonymous namespace)::thread_proxy (param=<optimized out>) at libs/thread/src/pthread/thread.cpp:171
#26 0x00007ffff7d75897 in start_thread (arg=<optimized out>) at pthread_create.c:444
#27 0x00007ffff7dfc6fc in clone3 () at ../sysdeps/unix/sysv/linux/x86_64/clone3.S:78 
```

Reply:
What was the git HEAD commit when you compiled this binary? I also had a segfault here in a previous commit, but I have pushed a fix (or so I thought) since then.

Reply:
The fix is working. Thanks.